### ISSUE TITLE
optimize.differential_evolution irregular convergence

### ISSUE THREAD

**Issue by timjim333** [scipy/scipy#13784]
Hi, not sure if this is a bug or perhaps I'm using it wrong, but I'm having some issues with optimize.differential_evolution. I'm using it to minimize a fitness function that will return `np.finfo("float").tiny` for certain variable combinations (infeasible). For some optimization cycles, a lot of designs will evaluate to something like `2.0140434892876102e-306`. The part I don't get is that sometimes, differential_evolution will continue to run until it finds a minimized solution (taking a few hundred iterations), yet other times it seems to give up just after the first iteration and call it a 'success', something like:

```
>> res = differential_evolution(func, optimbound, init=init_seed, args=args, workers=n_cpu, disp=True,
                                popsize=15, tol=1e-2, mutation=(0.8, 1.2), recombination=0.5, polish=True)
>> print(f'success: {res.success}, it: {res.get("nfev", None)}, msg: {res.message}')
... (tiny numbers here)
1.9052392031551207e-306
1.544023009525937e-306
1.4608397231988747e-307
differential_evolution step 1: f(x)= 0
7.992347617112624e-307
1.1456944063218226e-306
... (more tiny numbers here)
1.2894639333914132e-306
4.065250951391086e-307
5.3397473624036565e-307
success: True, it: 683, msg: Optimization terminated successfully.
```

Perhaps I'm misunderstanding the tolerance or convergence settings?
I'd appreciate any thoughts on this, and please let me know if I can provide more information! Thank you.

EDIT:
Just to add - does some kind of float casting go on inside differential_evolution? Although our fitness function explicitly filters out zeros and replaces them with tiny numbers, it seems like the result `res['fun']` is sometimes 0.0! 

#### Scipy/Numpy/Python version information:
1.5.2 1.19.2 sys.version_info(major=3, minor=7, micro=8, releaselevel='final', serial=0)

**Comment by andyfaff** [comment]
If a solution is infeasible then your fitness function should be returning `np.inf`.

Is the true value of your fitness function on the order of 1e-306? In general working with those kinds of numbers is not advisable as there will be next to no precision. You need to think about the scale of your fitness function. It's hard to advise without further details of what func is.

 The solve process `stops when ``np.std(pop) <= atol + tol * np.abs(np.mean(population_energies))``, where and `atol` and `tol` are the absolute and relative tolerance respectively.`. Normally `atol` is 0. If the fitness function values are all around 0 (and 1e-306 is to all intents and purposes 0), then that condition will be satisfied almost straight away.

There is no float casting of your fitness function, unless your fitness function is not returning double precision numbers.

**Comment by timjim333** [comment]
Hi @andyfaff thanks for the reply. Apologies, I worded the above a bit poorly.

The fitness function is an evaluation of the negative of the expected hypervolume improvement of the design space and is part of a larger framework of code, so I don't think I can edit it (or maybe I can modify my local copy for use with DE).

I was wrong when I said infeasible, I should have said there is zero improvement of the hypervolume from that given sample. In this case, the function appears to return a tiny positive number (as set by `np.finfo("float").tiny`) rather than zero exactly to avoid errors in another solver that is used. However, for samples that are an improvement, the truth value, as you say, should evaluate to some negative value, the more negative the better.

The search space is quite difficult, so I imagine that occasionally, nearly all of the sampled options may result in these tiny fitness values... What would be better? Would replacing all  `np.finfo("float").tiny` values with  `np.inf` fix this issue? I'm assuming  `population_energies` refers to the group fitness? In which case, does this just make all of `atol + tol * np.abs(np.mean(population_energies)) = 0`?

Thanks for your thoughts.

**Comment by andyfaff** [comment]
I think you might have something like this?
![Untitled](https://user-images.githubusercontent.com/702934/113460061-2d713880-9463-11eb-95ac-661d969b1292.png)

The graph shows the variation of a 1-D fitness function as a function of its value. Normally most optimisers would use the unclipped version (< 0 for a hypervolume improvement, > 0 for when it's degrading). What I think your function is doing is clipping it at *TINY* if the hypervolume doesn't improve, this would be the straight line. When the `differential_evolution` starts off if it doesn't find any initial parameter vectors with a fitness less than 0 then it will probably halt immediately, this is a consequence of clipping at 0. To combat this I suggest you make the values that don't improve `np.inf`.

`differential_evolution` holds a population of parameter vectors. The `population_energies` correspond to the fitness for each of the parameter vectors. The rationale for the halting tolerance is when there is little variation in population energies left. Your initial setup makes it highly likely that there will be little variation to start with, as everything is clipped.

**Comment by timjim333** [comment]
Hi, @andyfaff - you are spot on. Thank you, let me try setting those to np.inf. I noticed that `differential_evolution` seems to evaluate the fitness of each individual of the population one at a time. The fitness function I've supplied uses numpy arrays, so it can actually produce a `population`-length array of fitness values for an input array of shape `[n_population, n_design_vars]`; is there any way to take advantage of this rather than cycling through the population individually?
Thanks again!

**Comment by timjim333** [comment]
The other thing that seems to be happening is:
```
differential_evolution step 1: f(x)= nan
differential_evolution step 2: f(x)= nan
differential_evolution step 3: f(x)= nan
...
```
Even though for each step, there are some of the values which are actually not np.inf. Is there a reason why it selects nan rather than valid values?

**Comment by andyfaff** [comment]
> The fitness function I've supplied uses numpy arrays, so it can actually produce a population-length array of fitness values for an input array of shape [n_population, n_design_vars]; is there any way to take advantage of this rather than cycling through the population individually?

At the moment there is no way of allowing the fitness function to be evaluated in a vectorised way. It might be possible for you to create something to give that effect with use of the 'deferred' update and a map-like `workers` function, but it would require a bit of fiddling around
For a subset of problems the 'deferred' update is less likely to be successful compared to 'immediate'; the latter is the default but is not compatible with vectorisation.

Is this issue able to be closed now?

**Comment by timjim333** [comment]
If it's ok, I'd like to let the test finish first - I've not received anything back but `nan`s currently...

**Comment by andyfaff** [comment]
> Even though for each step, there are some of the values which are actually not np.inf. Is there a reason why it selects nan rather than valid values?

Not easy to forecast unless you can provide a MWE (minimal working example).

**Comment by timjim333** [comment]
I'll try to figure out how I can extract that section of things in a working MWE.

**Comment by timjim333** [comment]
In the event of non-convergence (hit max iter), what sample of the population does `differential_evolution` pick to return? The best-minimised sample so far? It's a bit unexpected but the returned sample doesn't seem to have been filtered for feasibility according to the upper and lower bounds. Is this a bug or intentional? Thanks.

**Comment by andyfaff** [comment]
`differential_evolution` will always return the best solution found up until that point, presuming there are some valid solutions found. The solutions should **always** lie in between the lower and upper bounds. 
If there is an upper bound equal to a lower bound somewhere it's possible that the LBFGSB polishing step may create problems.

**Comment by timjim333** [comment]
Sorry for the slow reply - it's been a bit mad , I've had a minute more to investigate the returned infeasible solutions.

Some of the constraints are evaluated as part of the fitness function, as the bounds are not truly independent; so the func = the HV improvement x feasibility due to extra constraints. 

When applying the np.inf edit for hypervolume improvements of 0, it seems that when the DE solver 'converges' for a fitness value of zero, the samples produced are also infeasible w.r.t. the constraints placed as part of the fitness function. This seems a little strange, however, as

**Comment by andyfaff** [comment]
If you need to apply constraints you should consider using the `constraints` keyword. This allows you to apply `NonLinearConstraint` or `LinearConstraint` to the modelling system.

**Comment by timjim333** [comment]
With the `constraints` keyword, would the aim be to move all design variable constraint functions there and have the fitness function only be driven by HV? I'll investigate this.

**Comment by timjim333** [comment]
Hi @andyfaff can I trouble you with a quick question about applying a 'dynamic' constraint bound? I understand that I can rephrase my constraints to something like:
```python
# Max rearward location at centreline
def n_5(x):
    """Max aft wing root location: x_offset + chord_0"""
    x_aft = x[0] + x[2]
    return x_aft

nlc_5 = so.NonlinearConstraint(n_5, -np.inf, X_AFT_MAX)

constraints = [nlc_1, nlc_2, ..., etc]
```
and collect up all the constraints to pass to `differential_evolution` but how can I specify a constraint with bounds that also depend on x? For instance, if `X_AFT_MAX` was determined by some other function `g(x)`?

**Comment by timjim333** [comment]
Or could I do the bound evaluation inside the constraint function and just return `np.nan` or some other sentinel value if it falls outside of my dynamic constraints? I'm not sure how `differential_evolution` is hooked up to handle this kind of thing.

**Comment by andyfaff** [comment]
A good example for a non linear constraint is at the [bottom of this page](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.NonlinearConstraint.html). The lower and upper bounds need to be numbers. In your example you could subtract X_AFT_MAX (possibly in the form of g(x)) in the last line of `n_5` and make the upper bound 0.

**Comment by timjim333** [comment]
Hi, @andyfaff thanks for your help and tips. Moving the constraints out of the fitness function and into `constraints`  with `NonLinearConstraint`s fixed the issue of infeasible samples appearing in the converged samples (even for 0 EHVI).

**Comment by timjim333** [comment]
Hi @andyfaff not sure if this belongs in a new topic, but thought I'd ask here quickly. Is it normal for `polish` to take quite a bit longer than the DE part of the optimization? For my current dataset, it's taking around 1.5 mins to do 500 generations (around 70k function calls) but it seems to be polishing for a good 8 mins or so (only an estimate from when `disp` stops showing the generations to the end of the run, so it could be other stuff going on there). I'm passing in `constraints`, so it's using the `trust-constr`  minimisation, is that correct? Thank you.

**Comment by timjim333** [comment]
While I remember, occasionally, I'm running across `ERROR:root:array must not contain infs or NaNs` while running DE, but this doesn't make much sense as I'm sure I've seen fitness of values np.infs being treated properly. Unfortunately, this tends to happen when `disp=False` so I don't know what triggers it.

**Comment by andyfaff** [comment]
> Hi not sure if this belongs in a new topic, 

GitHub issues is meant mainly for bug reports, the scipy mailing list is better for usage issues.

Normally the polish step is quite fast, but it will be system dependent. When there are constraints involved the `trust-constr` minimize method is used (otherwise `L-BFGS-B`). The speed of polishing will be determined by `trust-constr` alone. I wouldn't expect a polishing step to get that much improvement (depending on your point of view) over the end point of the differential evolution.

What you can try to do is do the `differential_evolution` with polishing turned off, then use `slsqp` in a separate `minimize` call. I'm not sure if that will be faster or not. 

Both minimize methods will have to calculate a function value (F), and a gradient (G). G is normally determined by finite differences, so take a minimum of `len(x) + 1` function evaluations themselves. F+G are also required for the constraints, so can suck up function evaluations as well.

**Comment by andyfaff** [comment]
> While I remember, occasionally, I'm running across ERROR:root:array must not contain infs or NaNs while running DE, but this doesn't make much sense as I'm sure I've seen fitness of values np.infs being treated properly. Unfortunately, this tends to happen when disp=False so I don't know what triggers it.

Is an Exception raised when this message is printed? What is the stack trace? Is that the exact error message that you see?  That message is not present in any of the differential evolution code, so is either coming from `trust-constr` somehow, or from your own code.

**Comment by timjim333** [comment]
Sorry, I should have given the full stack trace. Yes, it seems to be from something that DE passes to `trust-constr` in the polishing step, I think:
```
Traceback (most recent call last):
  File "mobo_B.py", line 123, in <module>
    xupdate, yupdate, supdate, metricall = optim.update_sample(update_info, n_kb)
  File "/work/projects/twb/2_OPTIMISATION/mobo_A.py", line 232, in update_sample
    xupdate, yupdate, supdate, metricall = mobo.run(disp=True, infeasible=None)
  File "/opt/kadal/KADAL/kadal/optim_tools/MOBO.py", line 105, in run
    self.ehviupdate(disp)
  File "/opt/kadal/KADAL/kadal/optim_tools/MOBO.py", line 156, in ehviupdate
    xnext, yprednext, sprednext, metricnext = self.simultpredehvi(disp)
  File "/opt/kadal/KADAL/kadal/optim_tools/MOBO.py", line 257, in simultpredehvi
    self.cheapconstlist)
  File "/opt/kadal/KADAL/kadal/optim_tools/acquifunc_opt.py", line 373, in run_multi_opt
    res = differential_evolution(*de_args, **de_kwargs, workers=pool.map)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_differentialevolution.py", line 308, in differential_evolution
    ret = solver.solve()
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_differentialevolution.py", line 814, in solve
    constraints=self.constraints)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_minimize.py", line 632, in minimize
    callback=callback, **options)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py", line 520, in _minimize_trustregion_constr
    factorization_method)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 328, in tr_interior_point
    factorization_method, trust_lb, trust_ub, subprob.scaling)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_trustregion_constr/equality_constrained_sqp.py", line 120, in equality_constrained_sqp
    lb_t, ub_t)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_trustregion_constr/qp_subproblem.py", line 498, in projected_cg
    r = Z.dot(H.dot(x) + c)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/sparse/linalg/interface.py", line 418, in dot
    return self.matvec(x)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/sparse/linalg/interface.py", line 232, in matvec
    y = self._matvec(x)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/sparse/linalg/interface.py", line 530, in _matvec
    return self.__matvec_impl(x)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/optimize/_trustregion_constr/projections.py", line 192, in null_space
    aux2 = scipy.linalg.solve_triangular(R, aux1, lower=False)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/linalg/basic.py", line 334, in solve_triangular
    b1 = _asarray_validated(b, check_finite=check_finite)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/scipy/_lib/_util.py", line 262, in _asarray_validated
    a = toarray(a)
  File "/home/user/anaconda3/envs/test-env/lib/python3.7/site-packages/numpy/lib/function_base.py", line 489, in asarray_chkfinite
    "array must not contain infs or NaNs")
ValueError: array must not contain infs or NaNs
```

**Comment by timjim333** [comment]
Hi @andyfaff does the trace help at all? Or is this a `minimize` issue? Thank you!

**Comment by andyfaff** [comment]
It's from trust-constr operation. Not sure it's an issue with the code or from your stuff. Hopefully by the time the polishing is done your code is returning finite numbers for the objective and constraint functions.

**Comment by timjim333** [comment]
This seems to happen when DE is passing the population to the polishing step. Does the `trust-constr` package follow your convention of assigning `np.inf` to infeasible members? If not, perhaps it's better to assign an arbitrarily large finite positive value than `np.inf`, if you are passing the population members directly to `minimize`?

**Comment by timjim333** [comment]
Potential bug - the `mp.Pool` supplied to DE gets closed once the optimiser is finished with it. Seems to be due to the `__exit__` function:
```
    def __exit__(self, *args):
        return self._mapwrapper.__exit__(*args)
```
which called when the `DifferentialEvolutionSolver` context manager ends. While closing the pool makes sense if it is instantiated by DE, it's a bit unexpected for an existing one that is passed in to be consumed. At least, my use case was to keep a pool open for several tight-loop tasks so that the expense of opening a new pool every iteration is reduced.

**Comment by andyfaff** [comment]
I don't think there is a bug there, `MapWrapper` is designed for that:

```
import numpy as np
from scipy.optimize import rosen, differential_evolution
import multiprocessing as mp

with mp.Pool(2) as p:
    res = differential_evolution(rosen, [(-2, 2)]*3, workers=p.map)
    print(p)
    print(list(p.map(rosen, [(1,1)]*10)))
```
results in:
```
<multiprocessing.pool.Pool state=RUN pool_size=2>
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```

**Comment by andyfaff** [comment]
In general nested parallelism is a bad idea.

**Comment by timjim333** [comment]
Thanks, it was another renegade function that was closing the pool passed to it. When you say nested parallelism, is that pool children that attempt to fire off their own pools? I would agree that this is a bad idea. Or do you mean for passing around an existing pool? Due to the cost of initialising a pool, it made sense to keep one open at a level above the tight loop, I thought.
